### PR TITLE
[megatron] Fix: fix bugs in mcore backend context-parallel code logic

### PR DESCRIPTION
**Pull request by Kite0011**
### What does this PR do?

This PR fixes two bugs related to CP in the mcore backend:

1. Fixed the label roll bug after CP selection, which caused loss misalignment before and after enabling CP
2. Fixed the loss scale bug when CP is enabled, which caused gradient misalignment before and after enabling CP, resulting in differences in loss convergence.

### Test

By fixing bug 1, the outlier in the logprob plot disappears. This point is the boundary point of CP splitting.

a means mcore with cp
b means mcore
c means fsdp
before fix logprob diff
image

after fix logprob diff
image

By fixing bug 2, the gradient norm can be aligned before and after enabling/disabling CP (Checkpoint).

Clipboard_Screenshot_1763991159

### Design & Code Changes

+ bug 1
Because the seq_pack preprocessing is currently quite complex, after completing the packing, we correctly rolled the labels based on the CP select logic.
+ bug 2
Since we have already performed a loss scale for all tokens within the DP in [sft_loss func](https://github.com/volcengine/verl/blob/main/verl/workers/roles/utils/losses.py#L48), there is no need to counteract the scaling at the CP level in mcore here.

### CODE REVIEW

File: verl/models/mcore/util.py [hunk 1]
```diff
@@ -228,6 +228,25 @@ def preprocess_packed_seqs_no_padding(
...
+                    input_ids_rmpad[start_idx + half_seqlen - 1] = d[(cp_rank + 1) * half_seqlen]
+                    remain_start = seqlen_padded_i - half_seqlen * (cp_rank + 1)
+                    remain_end = seqlen_padded_i - half_seqlen * cp_rank
+                    remain_end = min(remain_end, d.shape[0])
+                    remain_len = remain_end - remain_start
+                    if remain_len > 0:
+                        if remain_end == d.shape[0]:
+                            input_ids_rmpad[start_idx + half_seqlen + remain_len - 1] = d[0]
+                        else:
+                            input_ids_rmpad[start_idx + half_seqlen + remain_len - 1] = d[remain_end]
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/models/mcore/util.py:248]: using dict instead of double loop resolve this problem

File: verl/models/mcore/util.py [hunk 2]
```diff
@@ -229,6 +231,21 @@ def preprocess_packed_seqs_no_padding(
...
+                    if remain_end == d.shape[0]:
+                        saved_roll_dict[start_idx + half_seqlen + remain_len - 1] = d[0]
+                    else:
+                        saved_roll_dict[start_idx + half_seqlen + remain_len - 1] = d[remain_end]
+
+        if need_roll:
+            input_ids_rmpad = torch.roll(input_ids_rmpad, shifts=-1, dims=0)
+            if len(saved_roll_dict) > 0:
+                for k, v in saved_roll_dict.items():
+                    input_ids_rmpad[k] = v
```
[the remark this answers is not in the document]
Comment (Kite0011) [reply verl/models/mcore/util.py:247]: In the cp=1 scenario, boundary correction is not needed because the boundary positions between batches after rolling do not require training.

**Review by vermouth1992 — APPROVED** [review]
